docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,386 @@
+        {
+            var tollCalc = new TollCalculator();
+
+            var car = new Car();
+            var taxi = new Taxi();
+            var bus = new Bus();
+            var truck = new DeliveryTruck();
+
+            Console.WriteLine($"The toll for a car is {tollCalc.CalculateToll(car)}");
+            Console.WriteLine($"The toll for a taxi is {tollCalc.CalculateToll(taxi)}");
+            Console.WriteLine($"The toll for a bus is {tollCalc.CalculateToll(bus)}");
+            Console.WriteLine($"The toll for a truck is {tollCalc.CalculateToll(truck)}");
+
+            try
+            {
+                tollCalc.CalculateToll("this will fail");
+            }
+            catch (ArgumentException e)
+            {
+                Console.WriteLine("Caught an argument exception when using the wrong type", DayOfWeek.Friday);
+            }
+            try
+            {
+                tollCalc.CalculateToll(null);
+            }
+            catch (ArgumentNullException e)
+            {
+                Console.WriteLine("Caught an argument exception when using null");
+            }
+        }
+    }
+}
+```
+
+That code is included in the starter project, but is commented out. Remove the comments, and you can test what you've written.
+
+You're starting to see how patterns can help you create algorithms where the code and the data are separate. The `switch` expression tests the type and produces different values based on the results. That's only the beginning.
+
+## Add occupancy pricing
+
+The toll authority wants to encourage vehicles to travel at maximum capacity. They've decided to charge more when vehicles have fewer passengers, and encourage full vehicles by offering lower pricing:
+
+- Cars and taxis with no passengers pay an extra $0.50.
+- Cars and taxis with two passengers get a 0.50 discount.
+- Cars and taxis with three or more passengers get a $1.00 discount.
+- Buses that are less than 50% full pay an extra $2.00.
+- Buses that are more than 90% full get a $1.00 discount.
+
+These rules can be implemented using the **property pattern** in the same switch expression. The property pattern examines properties of the object once the type has been determined.  The single case for a `Car` expands to four different cases:
+
+```csharp
+vehicle switch
+{
+    Car { Passengers: 0} => 2.00m + 0.50m,
+    Car { Passengers: 1 } => 2.0m,
+    Car { Passengers: 2} => 2.0m - 0.50m,
+    Car c when c.Passengers > 2 => 2.00m - 1.0m,
+
+    // ...
+};
+```
+
+The first three cases test the type as a `Car`, then check the value of the `Passengers` property. If both match, that expression is evaluated and returned. The final clause shows the `when` clause for a property pattern. You use the `when` clause to test conditions other than equality on a property. In the preceding example, the `when` clause tests to see that there are more than 2 passengers in the car. Strictly speaking, it's not necessary in this example.
```

Comment:
Last sentence: 

The when clause illustrates the technique, but it's not necessary in this case because all other expected cases are accounted for.

(Its not just not needed, I think it doesn't help the clarity).

Reply:
Fixed, and moved the discussion of `when` to the bus switch arms, where it is used in a real condition.